### PR TITLE
make: fix sign-compare errors

### PR DESCRIPTION
**Pull request by smlng**
PR is required to achieve #7919, and contains several fixes for sign-compare errors

### CODE REVIEW

**Comment by smlng** [comment]
more to come, here too

**Comment by miri64** [comment]
Is this still WIP?

**Comment by smlng** [comment]
> Is this still WIP?

I think I just fixed the last thing missing to get #7919 ready to merge, so this and #7993 and #7995 are not WIP anymore and all 4 should be merged swiftly before anything/one breaks "-Wall -Wextra" again.

**Comment by miri64** [comment]
Test post incoming. Sorry this is the best PR I could think of :-/.

**Comment by miri64** [comment]
\o/

**Comment by miri64** [comment]
The thingy is now live so this comment might show up once again in this PR, but after that it shouldn't anymore ;-).

File: cpu/atmega_common/periph/i2c.c [hunk 1]
```diff
@@ -51,7 +51,7 @@ static mutex_t lock = MUTEX_INIT;
 int i2c_init_master(i2c_t dev, i2c_speed_t speed)
 {
     /* TWI Bit Rate Register - division factor for the bit rate generator */
-    uint8_t twibrr;
+    unsigned long  twibrr;
```
Comment (cladmi) [inline cpu/atmega_common/periph/i2c.c]: Extra whitespace added here.

File: cpu/atmega_common/periph/i2c.c [hunk 2]
```diff
@@ -112,7 +114,7 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
     /* disable device */
     TWCR &= ~(1 << TWEN);
     /* configure I2C clock */
-    TWBR = twibrr;                // Set TWI Bit Rate Register
+    TWBR = (uint8_t)twibrr;                // Set TWI Bit Rate Register
```
Comment (cladmi) [inline cpu/atmega_common/periph/i2c.c]: Messed up the comment alignment.

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 3]
```diff
@@ -256,22 +256,18 @@ void gnrc_ipv6_nib_handle_timer_event(void *ctx, uint16_t type)
...
-    }
-
-    if (try_long && ((res = gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG, 0,
-                                            l2src, l2src_maxlen)) > max_short_len)) {
+    /* get source address length */
+    int res = gnrc_netapi_get(iface, NETOPT_SRC_LEN, 0,
+                              &l2src_len, sizeof(l2src_len));
+    /* get source address, try long address first if possible  */
+    if ((res >= 0) && (l2src_len > max_short_len) &&
+        (gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG, 0,
```
Comment (cladmi) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c]: Here you changed the behavior, the previous code, set `res` to the result of `gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG,..)`, here it gets the result of `gnrc_netapi_get(iface, NETOPT_SRC_LEN,..)`.

File: sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c [hunk 4]
```diff
@@ -539,22 +539,18 @@ static gnrc_pktsnip_t *_build_headers(gnrc_ipv6_netif_t *netif,
...
-
-    if (try_long && ((res = gnrc_netapi_get(netif->pid, NETOPT_ADDRESS_LONG, 0,
-                                            l2src,
-                                            l2src_maxlen)) > max_short_len)) {
+    /* get source address length */
+    int res = gnrc_netapi_get(netif->pid, NETOPT_SRC_LEN, 0,
+                              &l2src_len, sizeof(l2src_len));
+    /* get source address, try long address first if possible  */
+    if ((res >= 0) && (l2src_len > max_short_len) &&
+        (gnrc_netapi_get(netif->pid, NETOPT_ADDRESS_LONG, 0,
```
Comment (cladmi) [inline sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c]: Same here you changed the behavior, the previous code, set res to the result of gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG,..), here it gets the result of gnrc_netapi_get(iface, NETOPT_SRC_LEN,..).

File: sys/posix/pthread/pthread_cond.c [hunk 5]
```diff
@@ -122,7 +122,7 @@ int pthread_cond_wait(pthread_cond_t *cond, mutex_t *mutex)
 
 int pthread_cond_timedwait(pthread_cond_t *cond, mutex_t *mutex, const struct timespec *abstime)
 {
-    if ((cond == NULL) || (abstime->tv_sec < 0)) {
+    if (cond == NULL) {
```
Comment (cladmi) [inline sys/posix/pthread/pthread_cond.c]: For some platforms `tv_sec` can be an int: https://github.com/RIOT-OS/RIOT/blob/master/cpu/atmega_common/include/sys/types.h#L38

So this is not equivalent to before.

Comment (smlng) [reply sys/posix/pthread/pthread_cond.c]: I know, but on some its `unsigned` - and that's the whole purpose of the struct: to hide actual implementation. See also [here](https://stackoverflow.com/questions/471248/what-is-ultimately-a-time-t-typedef-to)

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c (hunk 3 above)
Comment (smlng) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: I changed the assignment of `res` (that's right) and I also removed `try_long`, *but* did I really changed the behaviour or result of the function? Please also note: `res` isn't used afterwards anyway, the return value of this function was and still is `l2src_len`.

Comment (cladmi) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: I cannot see this change anymore, did you remove it ?
What I see on the original code, there is `l2src_len = (uint16_t)res;` just after this patch and so the return value would have taken the value of `res = gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG,)` which is not the case anymore.

File: sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c [hunk 6]
```diff
@@ -820,7 +820,7 @@ static size_t _get_l2src(kernel_pid_t iface, uint8_t *l2src, size_t l2src_maxlen
     }
 
     if (try_long && ((res = gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG, 0,
-                                            l2src, l2src_maxlen)) > max_short_len)) {
+                                            l2src, l2src_maxlen)) > (int)max_short_len)) {
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: I think without this change this wouldn't collide with #7925. This file is removed in that PR, so if you revert this change it could get in even with the GNRC embargo.

File: sys/shell/commands/sc_netif.c [hunk 7]
```diff
@@ -1295,7 +1296,7 @@ int _netif_config(int argc, char **argv)
                 }
                 int hl;
                 gnrc_ipv6_netif_t *entry;
-                if (((hl = atoi(argv[3])) < 0) || (hl > UINT8_MAX)) {
+                if (((hl = atoi(argv[3])) < 0) || (hl > (int)UINT8_MAX)) {
```
Comment (miri64) [inline sys/shell/commands/sc_netif.c]: Same goes for these lines.

File: sys/posix/pthread/pthread_cond.c (hunk 5 above)
Comment (cladmi) [reply sys/posix/pthread/pthread_cond.c]: For platform with signed `tv_sec`, the evaluation of 'then` would be wrong with negative values.
So maybe it should work only with `struct timespec` for comparing.

**Comment by smlng** [comment]
@miri64, I would still wait until #7925 is merged and rebase afterwards - because removing the changes here makes this PR inconsistent when used on its own with current master.

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c (hunk 3 above)
Comment (smlng) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: badly rebased, and thereby dropped some commit, I reintroduced them - so you should see them again. But also: I still don't think I changed the behaviour of the function - just reorganised and optimised the code a bit.

File: sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c (hunk 4 above)
Comment (smlng) [reply sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c]: seems to be copy-pasta from above - so my answer applies here too

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c (hunk 3 above)
Comment (cladmi) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: The last line should be `if ((res >= 0) && (l2src_len > max_short_len) && (res = gnrc_net_api_get(iface,NETOPT_ADDRESS_LONG, ...) > (int)max_short_len))` with `res =` to be the same behaviour as the next line is `l2src_len = (uint16_t)res;`

Comment (smlng) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: you're right, fixed - below too

File: sys/posix/pthread/pthread_cond.c (hunk 5 above)
Comment (cladmi) [reply sys/posix/pthread/pthread_cond.c]: This file is also missing now.
I planned to ping the person who added this `abstime->tv_sec < 0` to know if it made sense to remove it, but as it was you, I think you know what you are doing :)

**Comment by miri64** [comment]
> @miri64, I would still wait until #7925 is merged and rebase afterwards - because removing the changes here makes this PR inconsistent when used on its own with current master.

Alright. Just wanted to help to get this PR faster along (but my hopes for #7925 getting merged today are still high ;-))

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 8]
```diff
@@ -256,26 +256,22 @@ void gnrc_ipv6_nib_handle_timer_event(void *ctx, uint16_t type)
...
-    if ((gnrc_netapi_get(iface, NETOPT_SRC_LEN, 0, &l2src_len,
-                         sizeof(l2src_len)) >= 0) &&
-        (l2src_len > max_short_len)) {
-        try_long = true;
-    }
-
-    if (try_long && ((res = gnrc_netapi_get(iface, NETOPT_ADDRESS_LONG, 0,
-                                            l2src, l2src_maxlen)) > max_short_len)) {
+    /* get source address length */
+    int res = gnrc_netapi_get(iface, NETOPT_SRC_LEN, 0,
```
Comment (cladmi) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c]: @smlng seems good to me

@miri64 I think the change has now the same behavior as before but it would be perfect if you could double check this. (the same change is also in `gnrc_ndp2.c`).

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: Why is the variable declaration moved down? IMHO (and I stress that this is not a convention of the project but my personal) it is way more readable and comprehensible if it is at the top of the block, like it was before.

None of this changes except the cast in l270 seem to be about sign-compare errors, but maybe I'm mistaken. I'm not a friend of "massive" code restructuring in separate PRs and would prefer to put everything that isn't about the sign-compare error into a separate PR.

File: sys/net/gnrc/netif2/gnrc_netif2.c [hunk 9]
```diff
@@ -128,8 +128,8 @@ int gnrc_netif2_get_from_netdev(gnrc_netif2_t *netif, gnrc_netapi_opt_t *opt)
                 ipv6_addr_t *tgt = opt->data;
 
                 res = 0;
-                for (unsigned i = 0;
-                     (res < opt->data_len) && (i < GNRC_NETIF2_IPV6_ADDRS_NUMOF);
+                for (unsigned i = 0; (res < (int)opt->data_len) &&
```
Comment (miri64) [inline sys/net/gnrc/netif2/gnrc_netif2.c]: Here and below: Maybe have the end-condution completely on its own line.

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c (hunk 8 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: (btw: this is also a change that was simplified in #7925, since `gnrc_netif(2)` allows me to access the L2-address directly).

Comment (smlng) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: regarding code restructuring: I agree (my bad, was carried away by the idea of optimising _just a little_ 😉 ), such should be done in separate PRs if anyway.

Lets see what's left of it after merging #7925 and rebasing this one.

**Comment by miri64** [comment]
Please rebase

**Comment by miri64** [comment]
(GNRC embargo is lifted)

**Comment by smlng** [comment]
rebased, may I squash?

**Comment by cladmi** [comment]
Please squash

**Comment by smlng** [comment]
rebased and squashed

File: cpu/cc2538/include/cc2538_sys_ctrl.h [hunk 10]
```diff
@@ -151,9 +151,10 @@ typedef struct {
 /**
  * @brief Compute the current system clock frequency based on the SYS_CTRL register states
  */
-#define sys_clock_freq() ((SYS_CTRL->cc2538_sys_ctrl_clk_ctrl.CLOCK_CTRLbits.OSC ? \
-                           RCOSC16M_FREQ : XOSC32M_FREQ) >> \
-                                SYS_CTRL->cc2538_sys_ctrl_clk_ctrl.CLOCK_CTRLbits.SYS_DIV )
+#define sys_clock_freq() ((uint32_t)\
```
Comment (miri64) [inline cpu/cc2538/include/cc2538_sys_ctrl.h]: Weird point to break here IMHO, but okay.

**Comment by miri64** [comment]
The complaint by Travis is valid, but unrelated (and due to version differences in cppcheck). Let's merge and fix that issue in a separate PR.

**Comment by miri64** [comment]
See #8164 for the Travis issue fix.

**Comment by cladmi** [comment]
Thanks, we are moving toward -Wall -Wextra -Werror by default :)
